pyigm/fN/fnmodel.py

```diff
@@ -0,0 +1,456 @@
+        neval : int, optional
+          Discretization parameter (10000)
+        cumul : bool, optional
+          Return a cumulative array? (False)
+
+        Returns
+        -------
+        lX : float
+          l(X) value
+        """
+        # Initial
+        if NHI_max == None:
+            NHI_max = 23.
+            infinity=True
+        else:
+            infinity=False
+
+        try:
+            nz = len(z)
+        except:
+            nz=1
+            z = np.array([z])
+
+        # Brute force (should be good to ~0.5%)
+        lgNHI = np.linspace(NHI_min,NHI_max,neval)  #NHI_min + (NHI_max-NHI_min)*np.arange(neval)/(neval-1.)
+        dlgN = lgNHI[1]-lgNHI[0]
+
+        # Evaluate f(N,X)
+        lgfNX = self.evaluate(lgNHI, z)
+
+        # Sum
+        lX = np.zeros(nz)
+        for ii in range(nz): 
+            lX[ii] = np.sum(10.**(lgfNX[:, ii]+lgNHI)) * dlgN * np.log(10.)
+        if cumul == True:
+            if nz > 1:  #; Have not modified this yet
+                raise ValueError('fN.model: Not ready for this model type %s' % self.mtype)
+            cum_sum = np.cumsum(10.**(lgfNX[:, ii]+lgNHI)) * dlgN * np.log(10.)
+
+        # Infinity?
+        if infinity is True:
+            # This is risky...
+            # Best to cut it off
+            neval2 = 1000L
+            lgNHI2 = NHI_max + (99.-NHI_max)*np.arange(neval2)/(neval2-1.)
+            dlgN = lgNHI2[1] - lgNHI2[0]
+            lgfNX = np.zeros((neval2, nz))
+            lX2 = np.zeros(nz)
+            for ii in range(nz):
+                lgfNX[:, ii] = self.evaluate(lgNHI2, z[ii]).flatten()
+                lX2[ii] = np.sum(10.**(lgfNX[:, ii]+lgNHI2)) * dlgN * np.log(10.)
+            lX = lX + lX2
+
+        # Return
+        if nz == 1:
+            lX = lX[0]
+        if cumul == True:
+            return lX, cum_sum, lgNHI
+        else:
+            return lX
+
+    def calculate_rhoHI(self, z, NHI_mnx, neval=10000, cumul=False,
+        H0=70.*u.km/(u.s*u.Mpc)):
+        """ Calculate rho_HI over an N_HI interval
+
+        Parameters
+        ----------
+        z : float
+          Redshift for evaluation
+        NHI_mnx : tuple
+          of floats minimum/maximum NHI values
+        neval : int, optional
+          Discretization parameter (10000)
+        cumul : bool, optional
+          Return a cumulative array?
+        H0 : float, optional
+          Hubble's constant with units [70 km/s/Mpc]
+
+        Returns
+        -------
+        rho_HI: float
+          rho_HI in units of Msun per comoving Mpc**3
+        """
+        # Initial 
+        NHI_min=NHI_mnx[0]
+        NHI_max=NHI_mnx[1]
+        try:
+            nz = len(z)
+        except:
+            nz=1
+            z = np.array([z])
+
+        # Brute force (should be good to ~0.5%)
+        lgNHI = NHI_min + (NHI_max-NHI_min)*np.arange(neval)/(neval-1.)
+        dlgN = lgNHI[1]-lgNHI[0]
+
+        # Evaluate f(N,X)
+        lgfNX = self.evaluate(lgNHI, z)
```

Comment:
Are we ok in calling this without optional "cosmo"? H0 is used in L260, but dxdz in evaluate may use different H0? Risk of inconsistency?